lib/segment/src/common/arc_atomic_ref_cell_iterator.rs

```diff
@@ -0,0 +1,52 @@
+use atomic_refcell::AtomicRefCell;
+use std::sync::Arc;
+
+/// Intermediate data structure intended to help returning an iterator
+/// which depends on some guarded internal object.
+///
+/// The main idea is that we create a special implementation of iterator, which holds
+/// all required temporary values, which guarantees that source data for the iterator
+/// will be available during the whole time of the iterator existence.
+///
+/// In this specific case implemented support for `AtomicRefCell` but this code can easily
+/// be modified to other structures like `Mutex`, `RwLock`, e.t.c.
+pub struct ArcAtomicRefCellIterator<T: ?Sized, I> {
+    _holder: Arc<AtomicRefCell<T>>,
+    iterator: I,
+}
+
+impl<'a, T: 'a + ?Sized, I> ArcAtomicRefCellIterator<T, I> {
+    pub fn new<F>(holder: Arc<AtomicRefCell<T>>, f: F) -> Self
+    where
+        F: FnOnce(&'a T) -> I + 'a,
+    {
+        // We want to express that it's safe to keep the iterator around for as long as the
+        // Arc is around. Unfortunately, we can't say this directly with lifetimes, because
+        // we have to return iterator which depends on temporary `.borrow()` value:
+        //
+        // ```
+        // let reference = holder.borrow();
+        // //                    ^^^^^^^^^^ - temporary value created
+        // return ref.iter()
+        // ```
+        //
+        // Rust does not like that.
+        // That is why we use unsafe `.as_ptr()` to bypass the borrow checker.
+        // This operation should be safe overall, once we are keeping actual source of data alive
+        // with `_holder` in our reference counter.
+        // Even if the all other handlers are dropped - iterator will stay valid due to the reference counter.
+        let reference = unsafe { holder.as_ptr().as_ref().unwrap() };
```

Comment:
have you tried validating this with `miri`?

Reply:
Maybe we can add miri to our pipeline if it is stable enough

Reply:
Tried running `miri` and got some unrelated errors which do not help to assess the soundness of this change.

Below us the completely details for the sake of history.

- `cargo miri test --all`

```
test collection_manager::collection_updater::tests::test_payload_ops ... error: unsupported operation: can't call foreign function: epoll_create1
   --> /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.8.0/src/sys/unix/selector/epoll.rs:34:9
    |
34  |         syscall!(epoll_create1(flag)).map(|ep| Selector {
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function: epoll_create1
    |
    = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
            
    = note: inside `mio::sys::unix::selector::epoll::Selector::new` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.8.0/src/sys/unix/mod.rs:7:28
    = note: inside `mio::poll::Poll::new` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.8.0/src/poll.rs:377:13
    = note: inside `tokio::io::driver::Driver::new` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/io/driver/mod.rs:116:20
    = note: inside `tokio::runtime::driver::create_io_stack` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/runtime/driver.rs:22:29
    = note: inside `tokio::runtime::driver::Driver::new` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/runtime/driver.rs:170:52
    = note: inside `tokio::runtime::Builder::build_basic_runtime` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/runtime/builder.rs:560:35
    = note: inside `tokio::runtime::Builder::build` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/runtime/builder.rs:516:36
note: inside `collection_manager::collection_updater::tests::test_payload_ops` at lib/collection/src/collection_manager/collection_updater.rs:227:9
   --> lib/collection/src/collection_manager/collection_updater.rs:227:9
    |
227 |         assert!(!res[0].payload.as_ref().unwrap().contains_key("color"));
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside closure at lib/collection/src/collection_manager/collection_updater.rs:151:11
   --> lib/collection/src/collection_manager/collection_updater.rs:151:11
    |
150 |       #[tokio::test]
    |       -------------- in this procedural macro expansion
151 |       async fn test_payload_ops() {
    |  ___________^
152 | |         let dir = TempDir::new("segment_dir").unwrap();
153 | |         let segments = build_test_holder(dir.path());
154 | |         let searcher = SimpleCollectionSearcher::new();
...   |
227 | |         assert!(!res[0].payload.as_ref().unwrap().contains_key("color"));
228 | |     }
    | |_____^
    = note: this error originates in the macro `syscall` (in Nightly builds, run with -Z macro-backtrace for more info)

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error

```

- `cargo miri run  --bin qdrant`

```
error: unsupported operation: `getcwd` not available when isolation is enabled
   --> /home/agourlay/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/os.rs:143:17
    |
143 |             if !libc::getcwd(ptr, buf.capacity()).is_null() {
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `getcwd` not available when isolation is enabled
    |
    = help: pass the flag `-Zmiri-disable-isolation` to disable isolation;
    = help: or pass `-Zmiri-isolation-error=warn` to configure Miri to return an error code from isolated operations (if supported for that operation) and continue with a warning
            
    = note: inside `std::sys::unix::os::getcwd` at /home/agourlay/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/os.rs:143:17
    = note: inside `std::env::current_dir` at /home/agourlay/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/env.rs:54:5
    = note: inside `config::FileSourceFile::find_file::<config::FileFormat>` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/config-0.13.0/src/file/source/file.rs:33:13
    = note: inside `<config::FileSourceFile as config::file::source::FileSource<config::FileFormat>>::resolve` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/config-0.13.0/src/file/source/file.rs:106:34
    = note: inside `<config::File<config::FileSourceFile, config::FileFormat> as config::Source>::collect` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/config-0.13.0/src/file/mod.rs:128:45
    = note: inside `<config::File<config::FileSourceFile, config::FileFormat> as config::Source>::collect_to` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/config-0.13.0/src/source.rs:21:9
    = note: inside `<[std::boxed::Box<(dyn config::Source + std::marker::Send + std::marker::Sync + 'static)>] as config::Source>::collect` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/config-0.13.0/src/source.rs:113:13
    = note: inside `<[std::boxed::Box<dyn config::Source + std::marker::Send + std::marker::Sync>] as config::Source>::collect_to` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/config-0.13.0/src/source.rs:21:9
    = note: inside `config::ConfigBuilder::<config::builder::DefaultState>::build_internal` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/config-0.13.0/src/builder.rs:277:9
    = note: inside `config::ConfigBuilder::<config::builder::DefaultState>::build` at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/config-0.13.0/src/builder.rs:245:9
note: inside `settings::Settings::new` at src/settings.rs:28:17
   --> src/settings.rs:28:17
    |
28  |           let s = Config::builder()
    |  _________________^
29  | |             // Start off by merging in the "default" configuration file
30  | |             .add_source(File::with_name("config/config"))
31  | |             // Add in the current environment file
...   |
40  | |             .add_source(Environment::with_prefix("QDRANT").separator("__"))
41  | |             .build()?;
    | |____________________^
note: inside `main` at src/main.rs:23:20
   --> src/main.rs:23:20
    |
23  |     let settings = Settings::new().expect("Can't read config.");
    |                    ^^^^^^^^^^^^^^^

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error

```

Reply:
It is likely that miri can't work with tokio. Have you tried `-Zmiri-disable-isolation` for the second error?

Reply:
```
agourlay@agourlay-thinkpad:~/Workspace/qdrant$ MIRIFLAGS="-Zmiri-disable-isolation" cargo miri run  --bin qdrant
warning: /home/agourlay/Workspace/qdrant/Cargo.toml: unused manifest key: package.doctest
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
     Running `/home/agourlay/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/cargo-miri target/miri/x86_64-unknown-linux-gnu/debug/qdrant`
warning: thread support is experimental and incomplete: weak memory effects are not emulated.

error: unsupported operation: can't call foreign function: epoll_create1
  --> /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.8.0/src/sys/unix/selector/epoll.rs:34:9
   |
34 |         syscall!(epoll_create1(flag)).map(|ep| Selector {
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function: epoll_create1
```